### ISSUE TITLE
Nvidia Specific Issue: Flashlight "glow" rendered dim

### ISSUE THREAD

**Issue by Polymega** [elishacloud/Silent-Hill-2-Enhancements#74]
_@AeroWidescreen tagging you on this one because it is indeed an issue only for Nvidia users._

@FrozenFish24 if you'd have any ideas with this one, in regards to an additional Nvidia-specific issue for the game. 

This came to my attention the other day while doing testing for an upcoming mod by Aero: The glow around the flashlight lens is dimmer than what it should be for Nvidia cards. The lens flare seems to be fine; just the immediate glow around the lens is rendering incorrectly. 

On Intel GPUs and AMD cards the glow looks correct/as bright as it should be. When using WineD3D for Nvidia cards the glow is also correct: 

![image](https://user-images.githubusercontent.com/27395260/50530992-80fcbb00-0ad1-11e9-94c0-13f26520360f.png)

[Quick Save save file](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2715911/sh2pcsave2e.zip) of the image above (back up any save files first).

Intel Frame Capture: http://www.igotaletter.com/temp/sh2pc/sh2pc_2018_12_28__18_54_58.zip

Unfortunately the lens glow/lens flares don't render very well in the Intel frame grab but they're still there in the capture: 

![image](https://user-images.githubusercontent.com/27395260/50531081-e604e080-0ad2-11e9-89eb-3ce8bb674d5e.png)

**Comment by FrozenFish24** [comment]
Nice sleuthing @Polymega

This seems to be the same issue as the missing 2D fog and afaict it's happening because on Nvidia cards using a negative `DEPTHBIAS` biases some fullscreen element behind the camera.

D3D8 has a `ZBIAS` render state that takes an integer from 0-16. D3D9 replaced this with `DEPTHBIAS` which takes a float. d3d8to9 converts between the two by multiplying the `ZBIAS` by -0.000005. So a potential fix is to bitwise negate the `ZBIAS` first so we wind up with a positive `DEPTHBIAS` after d3d8to9 does it's thing:

```
HRESULT m_IDirect3DDevice8::SetRenderState(D3DRENDERSTATETYPE State, DWORD Value)
{
	if (State == D3DRS_ZBIAS)
		Value = !Value;

	return ProxyInterface->SetRenderState(State, Value);
}
```
[sh2e-dll.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2717209/sh2e-dll.zip)

This seems to work fine after some quick testing and is also an alternate fix for the missing 2D fog. But we should get @elishacloud's opinion on this as I recall he tried messing with the `ZBIAS` initially but abandoned it as it caused other issues.

**Comment by Polymega** [comment]
Through quick testing on my end it seems to have worked without breaking anything else (although a more thorough playtest would be required). 

I even disabled `d3d8to9` and `Fog2DFix` in the .ini and everything was still rendering correctly, which tells me those two fixes weren't influencing your fix for the flashlight glow (and the alternate fix solution for the 2D fog layer). 

If you and Elisha would like, I can do a complete playtest of this at a later time. Or if the above-mentioned fix isn't a proper method to alleviate the issue just let me know and I'll hold off.

**Comment by FrozenFish24** [comment]
Okay looking at it again with fresh eyes, my previous method (Even if I'd actually done it properly :stuck_out_tongue: ) wasn't a good idea as it's D3D9 dependent. Here's another idea:
```
HRESULT m_IDirect3DDevice8::SetRenderState(D3DRENDERSTATETYPE State, DWORD Value)
{
	if (State == D3DRS_ZBIAS)
		Value = (Value * 15) / 16;

	return ProxyInterface->SetRenderState(State, Value);
}
```

Basically we're taking the original `ZBIAS` value's range of 0-16 and squashing it down to a range of 0-15. So this fix may result in slightly more Z-fighting when enabled. This fixes both the flash light and the 2D fog and doesn't appear to introduce any new problems from a quick playtest.

Test dll:
[sh2e-dll.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2718027/sh2e-dll.zip)

Again we should get @elishacloud's feedback on whether it's a good idea to mess with the `ZBIAS` like this.

**Comment by Polymega** [comment]
> This fixes both the flash light and the 2D fog and doesn't appear to introduce any new problems from a quick playtest. Test dll:

Same here; a few quick tests of areas where I'd presume there'd be depth fighting there is none. :) We'll get Elisha's feedback and/or blessing on this, but thank you very much FrozenFish24 for your testing and iterations on this solution!

**Comment by elishacloud** [comment]
Nice work @FrozenFish24!  This looks like a generic fix for the layering issue with Nvidia cards.  I updated the `Fog2DFix` option to use this `ZBIAS` fix and removed the old fix.  Below is the updated build.  @Polymega, once you done enough testing I will check it in.

Updated build: [d3d8.zip](https://github.com/elishacloud/Silent-Hill-2-Enhancements/files/2719636/d3d8.zip)

**Comment by Polymega** [comment]
@elishacloud I finished the play test with the new depth fix and everything looked great! I saw no anomalies from this anywhere.

Additionally, I'm not sure if you confirmed, but your two vertex shader fix codes also worked without any issues found [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/63#issuecomment-444701621) and [here](https://github.com/elishacloud/Silent-Hill-2-Enhancements/issues/63#issuecomment-444727997). 

Thanks!

**Comment by elishacloud** [comment]
Checked in fix for this under [9ae79b7](https://github.com/elishacloud/Silent-Hill-2-Enhancements/commit/9ae79b7673e62978f3902e72aeb084edc9b1c676).  I'll fix the vertex shaders in issue #63.

**Comment by elishacloud** [comment]
@Polymega, I think this fix is going to take more than just changing the code of one pixel shader.

[MSDN says:](https://docs.microsoft.com/en-us/windows/desktop/direct3dhlsl/dx-graphics-hlsl-writing-shaders-9#pixel-shader-basics)

> Pixel shaders work in concert with vertex shaders; the output of a vertex shader provides the inputs for a pixel shader.

This probably requires you to create a vertex shader that targets only the window and then target the pixel shader to that vertex only.  This will require API changes in addition to shader changes.  Likely it will require at least one new shader to be create, rather than just modifying existing ones.
